### PR TITLE
fix(mocha framework): Select tests based on name

### LINKED ISSUE #249: Small variation in mutation score between coverageAnalysis perTest and all
There is a small variation in mutation score between `coverageAnalysis` `'perTest'` and `'all'`. I cannot explain this right away, we need more investigation.

On the master branch:

**Per test:**
```
1591 total mutants.
603 mutants survived.
  of which 288 were not covered by the tests.
4 mutant(s) caused an error and were therefore not accounted for in the mutation score.
47 mutants timed out.
937 mutants killed.
Ran 27.32 tests per mutant on average.
Mutation score based on covered code: 75.75%
Mutation score based on all code: 62.00%
[2017-03-02 07:54:47.168] [INFO] Stryker - Done in 2 minutes 29 seconds.
```

**all:**
```
1591 total mutants.
595 mutants survived.
  of which 288 were not covered by the tests.
4 mutant(s) caused an error and were therefore not accounted for in the mutation score.
47 mutants timed out.
945 mutants killed.
Ran 108.93 tests per mutant on average.
Mutation score based on covered code: 76.37%
Mutation score based on all code: 62.51%
```

On the "feat-baseline-coverage" branch:
**Per test:**
```
1624 total mutants.
617 mutants survived.
  of which 296 were not covered by the tests.
4 mutant(s) caused an error and were therefore not accounted for in the mutation score.
47 mutants timed out.
956 mutants killed.
Ran 48.65 tests per mutant on average.
Mutation score based on covered code: 75.76%
Mutation score based on all code: 61.91%
[2017-03-02 07:39:22.690] [INFO] Stryker - Done in 2 minutes 41 seconds.
```

**all:**
```
1624 total mutants.
611 mutants survived.
  of which 296 were not covered by the tests.
4 mutant(s) caused an error and were therefore not accounted for in the mutation score.
47 mutants timed out.
962 mutants killed.
Ran 110.40 tests per mutant on average.
Mutation score based on covered code: 76.21%
Mutation score based on all code: 62.28%
[2017-03-02 07:35:56.898] [INFO] Stryker - Done in 3 minutes 39 seconds.
```

### PR DESCRIPTION
**Pull request by nicojs**
Select tests in the `MochaTestFramework` based on it's name, rather than the order in which it is added to a test suite.

* Change TestFramework interface to now filter based on test `id` and `name`.
* Add `TestSelection` which represents a selected test using `id` and `name`.
* Update `JasmineTestFramework` to now use the new `TestSelection` format (still based on ID).
* Update `MochaTestFramework` to use the title to select tests.
* Add integration tests to both `JasmineTestFramework` and `MochaTestFramework` to prevent this sort of thing from happening in the future.

Fixes #249

BREAKING CHANGES:

* Change api of `TestFramework`. It now provides an array of `TestSelection` objects, instead of an array of numbers with test ids.

### CODE REVIEW

**Comment by riezebosch** [comment]
Probably also fixes https://github.com/stryker-mutator/stryker/issues/291

File: packages/stryker-jasmine/test/integration/nestedSuiteSpec.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    filter([2]);
+    const result = execJasmine(selectTestFile, nestedSuiteFile);
+    expect(result).lengthOf(3);
+    expect(result[0].status).eq('disabled');
+    expect(result[1].status).eq('disabled');
+    expect(result[2].status).eq('passed');
+    expect(result[2].fullName).eq('outer test 3');
+  });
+
+  function filter(testIds: number[]) {
```
Comment (simondel) [inline packages/stryker-jasmine/test/integration/nestedSuiteSpec.ts]: Do you ever pass anything other than a single number? If you don't won't it be easier to just allow one test id?

File: packages/stryker-mocha-framework/src/MochaTestFramework.ts [hunk 2]
```diff
@@ -14,22 +14,24 @@ export default class MochaTestFramework implements TestFramework {
...
-        if (${JSON.stringify(testIds)}.indexOf(current) > -1) {
+      var realAddTest = Mocha.Suite.prototype.addTest;
+      
+      Mocha.Suite.prototype.addTest = function (test) {
+        // Only actually add the tests with the expected names
+        var name = this.fullTitle() + ' ' + test.title;
+        if(selectedTestNames.indexOf(name) !== -1) {
           realAddTest.apply(this, arguments);
         }
-        current++;
```
Comment (simondel) [inline packages/stryker-mocha-framework/src/MochaTestFramework.ts]: You've removed this line but the declaration is still there. We can remove it right?

File: packages/stryker-mocha-framework/test/integration/nestedSuiteSpec.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+  it('should only run test 3 if filtered on index 2', () => {
+    filter([2]);
+    const result = execMocha(selectTestFile, nestedSuiteFile);
+    expect(result.tests).lengthOf(1);
+    expect(result.passes).lengthOf(1);
+    expect(result.passes[0].fullTitle).eq('outer inner test 3');
+  });
+
+  function filter(testIds: number[]) {
```
Comment (simondel) [inline packages/stryker-mocha-framework/test/integration/nestedSuiteSpec.ts]: The same filter comment for the Jasmine test applies here

File: packages/stryker-mocha-framework/src/MochaTestFramework.ts (hunk 2 above)
Comment (nicojs) [reply packages/stryker-mocha-framework/src/MochaTestFramework.ts]: Good point. I'll remove it.

File: packages/stryker-jasmine/test/integration/nestedSuiteSpec.ts (hunk 1 above)
Comment (nicojs) [reply packages/stryker-jasmine/test/integration/nestedSuiteSpec.ts]: Good point. I've added a test where we filter 2 tests just to make sure it works.

File: packages/stryker-mocha-framework/test/integration/nestedSuiteSpec.ts (hunk 3 above)
Comment (nicojs) [reply packages/stryker-mocha-framework/test/integration/nestedSuiteSpec.ts]: Good point. I've added a test where we filter 2 tests just to make sure it works.

**Pull request merged**